### ISSUE TITLE
Analyzing issues with Player Zero (2/2)

### ISSUE THREAD

**Issue by Klazkin** [Klazkin/player-zero#88]
Explain the context of the issue, what is being addressed in detail.

## The goal
- [x] Null loss after long training period
- [x] Verify that MCTS computed scores add up to 1.0
- [x] Simplify the policy mask logic
- [x] Look into the cause of large policy outputs.
- [x] Apply a proper policy mask in inference
- [x] Fix board representations life-time bug.
- [x] Fix policy vector life-time.
- [x] Look into differences between floats in python and c++
- [x] Investigate using higher precision floats.
- [x] Verify C++ de-serialization and Python de-serialization produce same results.
- [x] Fix Policy layer sum greatly exceeding 1.0
- [x] Check Expected policy in training server is correctly indexed
- [x] Check policy in self play data gen is index correctly
- [x] Hard-coded masking of softmax layer
- [x] (manually) Check where high policy loss comes from
- [x] Check value (winner)output
- [x] Add Conv2d layer kernel regularizers (https://github.com/NeymarL/ChineseChess-AlphaZero/blob/distributed/cchess_alphazero/agent/model.py)
- [ ] Auto learning rate https://github.com/NeymarL/ChineseChess-AlphaZero/blob/distributed/cchess_alphazero/worker/optimize.py#L138
- [x] alternative loss algorithms https://github.com/monkeyDemon/AI-Toolbox/blob/master/computer_vision/image_classification_keras/loss_function/focal_loss.py
- [x] https://github.com/Klazkin/gaf-root/issues/87

## Time tracking

Time Estimate: `undefined`
Time spent: `17 hours 10 minutes`

## Resources
https://tmoer.github.io/AlphaZero/
https://keras.io/api/layers/activation_layers/softmax/
https://stackoverflow.com/questions/56218256/nan-values-in-loss-in-keras-model
https://stackoverflow.com/questions/65131391/what-exactly-is-kerass-categoricalcrossentropy-doing
https://stackoverflow.com/questions/33962226/common-causes-of-nans-during-training-of-neural-networks
https://www.tensorflow.org/api_docs/python/tf/keras/Regularizer
https://datascience.stackexchange.com/questions/38358/dealing-with-extreme-values-in-softmax-cross-entropy
https://stackoverflow.com/questions/40050397/deep-learning-nan-loss-reasons
https://github.com/keras-team/keras/blob/f77b020e497a353b644df3aeebc97c831c8057fc/keras/layers/activations/softmax.py#L51
https://keras.io/api/layers/initializers/
